core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -789,6 +811,23 @@ private[spark] class ExecutorAllocationManager(
       }
     }
 
+    override def onUnschedulableBlacklistTaskSubmitted
+      (blacklistedTask: SparkListenerUnschedulableBlacklistTaskSubmitted): Unit = {
+      val stageId = blacklistedTask.stageId
+      val stageAttemptId = blacklistedTask.stageAttemptId
+      allocationManager.synchronized {
+        (stageId, stageAttemptId) match {
+          case (Some(stageId), Some(stageAttemptId)) =>
+            val stageAttempt = StageAttempt(stageId, stageAttemptId)
+            unschedulableTaskSets.add(stageAttempt)
+          case (None, _) =>
+            // Clear unschedulableTaskSets since atleast one task becomes schedulable now
+            unschedulableTaskSets.clear()
```

Comment:
Why are we clearing all tasksets ? Shouldn't this not remove the specific taskset which became schedulable ?

Reply:
https://github.com/apache/spark/blob/263f04db865920d9c10251517b00a1b477b58ff1/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L670 - Based on the previous implementation, we want to defer killing taskSets as long as we have a non blacklisted executor which can be used to schedule a task. The problem is if we can only schedule only one of the task set then the other taskset might expire and the stage gets aborted. This is the reason I think why they cleared the expiry in the previous implementation which is why we are also clearing the `unschedulableTaskSets`

Reply:
There is a difference in both cases.
If all tasksets are removed from `TaskSchedulerImpl.unschedulableTaskSetToExpiryTime`, it would defer aborting unschedulable tasksets in the assumption that some progress is being made and eventually tasksets might become scheduleable (or gets readded subsequently).
On other hand, if an unschedulable taskset remains in `unschedulableTaskSets`, this can trigger a successful executor allocation sooner - which could make it (and other taskset's) schedulable.

If we clear `unschedulableTaskSets` here, an eventual resourceOffer will result in re-addition to `unschedulableTaskSets` after a delay. If we want to defer resource acquisition, that sounds fine - even with this, current PR is still a strict improvement over what we have.

Reply:
Yeah I could go either way on this as well. if we only remove the task set it could help get more faster if lots of blacklisted or you only requested a few to begin with. I think its a big out of the norm for us to rely on the parameters here being None as well, traditionally we would have 2 messages, 1 for it being added, one for it being removed.  perhaps onUnschedulableTaskSetAdded and onUnschedulableTaskSetRemoved.